devices/raspizerow/I2s/main.cpp

```diff
@@ -0,0 +1,301 @@
+    file.write(reinterpret_cast<const char *>(&subchunk1Size), 4);
+    short audioFormat = 1; // PCM
+    file.write(reinterpret_cast<const char *>(&audioFormat), 2);
+    file.write(reinterpret_cast<const char *>(&channels), 2);
+    file.write(reinterpret_cast<const char *>(&sampleRate), 4);
+    int byteRate = sampleRate * channels * bitsPerSample / 8;
+    file.write(reinterpret_cast<const char *>(&byteRate), 4);
+    short blockAlign = channels * bitsPerSample / 8;
+    file.write(reinterpret_cast<const char *>(&blockAlign), 2);
+    file.write(reinterpret_cast<const char *>(&bitsPerSample), 2);
+    file.write("data", 4);
+    file.write(reinterpret_cast<const char *>(&dataSize), 4);
+}
+
+void recordAudio(snd_pcm_t *capture_handle, snd_pcm_uframes_t period_size)
+{
+    std::vector<char> buffer(period_size * bytesPerSample);
+    std::vector<char> accumulatedBuffer;
+
+    while (running)
+    {
+        rc = snd_pcm_readi(capture_handle, buffer.data(), period_size);
+        if (rc == -EPIPE)
+        {
+            // Handle overrun
+            std::cerr << "Overrun occurred" << std::endl;
+            snd_pcm_prepare(capture_handle);
+        }
+        else if (rc < 0)
+        {
+            std::cerr << "Error from read: " << snd_strerror(rc) << std::endl;
+            break; // Exit the loop on error
+        }
+        else if (rc != (int)period_size)
+        {
+            std::cerr << "Short read, read " << rc << " frames" << std::endl;
+        }
+        else
+        {
+            // Append the captured data to the accumulated buffer
+            accumulatedBuffer.insert(accumulatedBuffer.end(), buffer.begin(), buffer.begin() + rc * bytesPerSample * channels);
+
+            if (accumulatedBuffer.size() >= targetBytes)
+            {
+                audioQueue.push(accumulatedBuffer);
+                accumulatedBuffer.clear();
+            }
+        }
+    }
+}
+
+void handleAudioBuffer()
+{
+    while (running)
+    {
+        std::vector<char> dataChunk;
+        // Accumulate enough data to form a complete WAV file
+        while (dataChunk.size() < targetBytes)
+        {
+            std::vector<char> buffer = audioQueue.pop();
+            dataChunk.insert(dataChunk.end(), buffer.begin(), buffer.end());
+        }
+
+        // Process and send the accumulated data
+        if (!dataChunk.empty())
+        {
+            std::ofstream wavFile("output.wav", std::ios::binary);
+            int bitsPerSample = 32;
+            int dataSize = dataChunk.size();
+            writeWavHeader(wavFile, bitsPerSample, dataSize);
+            wavFile.write(dataChunk.data(), dataSize);
+            wavFile.close();
+
+            // Send the WAV file to the server
+            sendWav("output.wav");
+
+            // Delete the file after sending
+            std::filesystem::remove("output.wav");
+        }
+    }
+
+    // After exiting the loop, check if there's any remaining data to be processed
```

Comment:
Are we ever exiting the loop of the handleAudioBuffer? in what scenario it gets to this loop?

Reply:
I added that to handle sending any data that hasn't met the 60sec threshold and after the program has been ended. So when we ctrl-c to quit the program we go back to check and make sure theres no data that still needs to be sent

Reply:
So the loop only ends when we quit the program

Reply:
I see - so lets delete it, it's not needed if we are always-on and running on the device without an SSH terminal on - you won't be able to ctrl-C anyway.